### PR TITLE
fix: compilers with different settings were not disinguished

### PR DESCRIPTION
**Pull request by antazoey**
### What I did

`__eq__` and `__hash__` in `Compiler` did not account for settings.
This is actually really important in `ape-vyper` right now, as it groups input selections based on optimization pragmas.

### How I did it

first stringify  settings dict in a sorted consistent way _without_ `outputSelection` in it. Now, only compiler settings like evmVersion or optimization settings affect the str.

then place that str in the hash

### How to verify it

when using `ape-vyper` 's 0.7 PR you can compile a contract with opimization pragma and it gets own Compiler entry separate from the others using the same version..

See test updates on this PR for a simpler example lol.

### Checklist

- [ ] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [ ] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

**Comment by antazoey** [comment]
damnit, this still doesnt work because contract type names are in the settings so now it is seen as new compiler when they arent

File: ethpm_types/source.py [hunk 1]
```diff
@@ -42,19 +43,43 @@ class Compiler(BaseModel):
...
+            elif isinstance(value, list):
+                return [sort_value(item) for item in value]
+            else:
+                return value
+
+        def sort_dict(_dict: Dict) -> Dict:
+            return {
+                k: sort_value(v)
+                for k, v in sorted(_dict.items(), key=lambda item: sort_value(item[1]))
+            }
```
Comment (mikeshultz) [inline ethpm_types/source.py:76]: thought: double sort_value() calls could be costly depending on what we're talking about here.  Though building a list of tuples first might also suck.  :shrug: 

thought: Wonder if this is useful elsewhere enough to get moved to a util.  Deterministic object hashing has come up for me before in the past.

**Review by mikeshultz — APPROVED** [review]
LGTM. Unnecessary non-blocking commentary inline.

File: ethpm_types/source.py (hunk 1 above)
Comment (antazoey) [reply ethpm_types/source.py:76]: i was thinking about util too.

another though i had for performance was just whitelisting values to include so we can just pass in a few keys that are common changes, like `evmVersion` and `optimizer`

Comment (antazoey) [reply ethpm_types/source.py]: OK - so this comment made me think so I tried to address it anyway!

* Made the util for hashing strings.
* Got rid of double call to sort_value, wasnt needed, but still needed sorting keys of first dict.
* But bigger perf, made eq only check optimizer and evmVersion. In the future, can revisit a way to make more customizable but this at least solves the problem and isnt as costly.

**Pull request merged**